### ISSUE TITLE
[BUG] Building Dockerfile locally fails

### ISSUE THREAD

**Issue by polarathene** [docker-mailserver/docker-mailserver#2023]
# Bug Report

When cloning the project and attempting to build a Docker image locally via `make build` it was failing during package install.

<details>
<summary>Earlier report (invalid)</summary>

**EDIT:** These findings below were mistaken. I was under the belief that the errors I found when disabling the `SHELL` directive was the actual build failures, but those were new errors due to running `bash` syntax with `sh`, thus misleading. **Actual cause** is related to `postfix` package post install failing. I will add a new comment with details.

## Context

Appears to be [introduced in early Jan with the migration](https://github.com/docker-mailserver/docker-mailserver/commit/189e5376cc91af4f8a6b1e0db7b8f20e53a932a1#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R38) by @aendeavor . Usage of `&>/dev/null` was the culprit.

Additionally, the actual error context was hidden due to:

https://github.com/docker-mailserver/docker-mailserver/blob/abdf681d0265fce41aa971dd6b8194c902a3c1f6/Dockerfile#L35

### What is affected by this bug?

Building the Dockerfile locally under latest versions of Docker.

The redirection behaviour is illustrated quite well [here](https://unix.stackexchange.com/a/119650), with the `&> /dev/null` syntax described [here](https://askubuntu.com/a/1145025) with a comment stating this is deprecated in favor of `> /dev/null 2>&1` which it is apparently equivalent to.

Interestingly, if that is the desired behaviour (_to hide stdout, and redirect stderr into stdout to be hidden as well_) that does work as intended with `>/dev/null 2>&1`, just not `&>/dev/null`.

</details>

### When does this occur?

Within the Dockerfile.

<details>
<summary>From original report, unrelated to the real bug, but may want to fix the inconsistency</summary>

There are:

- 16 occurrences of `/dev/null` at present, 15 will match `>/dev/null`, while the other is inconsistent with a space `> /dev/null`. 6 use `&>/dev/null`.
- 4 occurrences of `2>&1`, but only 1 paired with `>/dev/null` but with a swapped order of `2>&1 >/dev/null`(_this redirects stderr into stdout, but original stdout output is hidden as it's sent to `/dev/null` instead_). Unclear if that was the intention.

</details>

### How do we replicate the issue?

1. Ubuntu 21.04 (I used Vultr if it makes any difference)
2. [Install Docker-Engine](https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository):

   ```bash
   apt-get update && apt-get install apt-transport-https
   curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg
   echo \
     "deb [arch=amd64 signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/ubuntu \
     $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
   
   apt-get update && apt-get install docker-ce docker-ce-cli containerd.io
   ```
3. Clone `docker-mailserver` and build image:

   ```bash
   git clone https://github.com/docker-mailserver/docker-mailserver.git
   cd docker-mailserver
   git submodule update --init --recursive
   apt-get install make
   make build
   ```
4. Build fails.

## Behavior

<details>
<summary>Earlier report (invalid)</summary>

With the first occurrence being changed to `>/dev/null` or `>/dev/null 2>&1`, packages install properly and there is no error about being unable to find packages such as `ca-certificates` (_It is unclear why `&>/dev/null` affected the earlier commands, but presumably it prevented them from running properly for some reason_).

<details>
<summary>Output</summary>

With the F2B section, as-is I noticed that some output was in red (stderr?):

```
gpg: directory '/root/.gnupg' created
gpg: keybox '/root/.gnupg/pubring.kbx' created                                                                                                                                            
```

and changing to `>/dev/null` retained those two lines, and additionally appended a few more (also in red):

```
gpg: /root/.gnupg/trustdb.gpg: trustdb created                                                                                                                                            
gpg: key 683BF1BEBD0A882C: public key "Serg G. Brester (sebres) <info@sebres.de>" imported                                                                                                
gpg: Total number processed: 1                                                                                                                                                            
gpg:               imported: 1  
```

The F2B `FINGERPRINT` script then uses `2>&1` 3 times, with some conditionals. The output as-is is:

```
/bin/sh: 1: [[: not found
/bin/sh: 1: [[: not found
```

When the `SHELL` directive is re-enabled, this is the output:
```
The command '/bin/bash -o pipefail -c dpkg -i fail2ban.deb &>/dev/null' returned a non-zero code: 1
```

---

Reduced Dockerfile used:

```Dockerfile
FROM docker.io/debian:buster-slim

ARG DEBIAN_FRONTEND=noninteractive

ARG FAIL2BAN_DEB_URL=https://github.com/fail2ban/fail2ban/releases/download/0.11.2/fail2ban_0.11.2-1.upstream1_all.deb
ARG FAIL2BAN_DEB_ASC_URL=${FAIL2BAN_DEB_URL}.asc
ARG FAIL2BAN_GPG_PUBLIC_KEY_ID=0x683BF1BEBD0A882C
ARG FAIL2BAN_GPG_PUBLIC_KEY_SERVER=keys.gnupg.net
ARG FAIL2BAN_GPG_FINGERPRINT="8738 559E 26F6 71DF 9E2C  6D9E 683B F1BE BD0A 882C"

#SHELL ["/bin/bash", "-o", "pipefail", "-c"]

RUN \
  apt-get -qq update && \
  apt-get -qq install apt-utils &>/dev/null && \
  apt-get -qq dist-upgrade >/dev/null
  
RUN apt-get -y --no-install-recommends install ca-certificates curl gnupg python3

RUN \
  # Fail2Ban
  gpg --keyserver ${FAIL2BAN_GPG_PUBLIC_KEY_SERVER} \
    --recv-keys ${FAIL2BAN_GPG_PUBLIC_KEY_ID} &>/dev/null && \
  curl -Lkso fail2ban.deb ${FAIL2BAN_DEB_URL} && \
  curl -Lkso fail2ban.deb.asc ${FAIL2BAN_DEB_ASC_URL}

RUN \
  FINGERPRINT="$(LANG=C gpg --verify \
  fail2ban.deb.asc fail2ban.deb 2>&1 \
    | sed -n 's#Primary key fingerprint: \(.*\)#\1#p')" && \
  if [[ -z ${FINGERPRINT} ]]; then \
    echo "ERROR: Invalid GPG signature!" 2>&1; exit 1; fi && \
  if [[ ${FINGERPRINT} != "${FAIL2BAN_GPG_FINGERPRINT}" ]]; then \
    echo "ERROR: Wrong GPG fingerprint!" 2>&1; exit 1; fi

RUN \
  dpkg -i fail2ban.deb &>/dev/null

RUN \
  rm fail2ban.deb fail2ban.deb.asc

RUN \
  # cleanup
  apt-get -qq autoremove &>/dev/null && \
  apt-get -qq autoclean && \
  apt-get -qq clean

RUN rm -rf /var/lib/apt/lists/*
RUN c_rehash &>/dev/null
```

```shell-session
$ docker build -t test-build -f Dockerfile .

...
Step 8/17 : RUN   apt-get -qq update &&   apt-get -y install apt-utils &>/dev/null &&   apt-get -y dist-upgrade >/dev/null
 ---> Running in f30ee26e1d1f
Removing intermediate container f30ee26e1d1f
 ---> 4fd0bfffc854
Step 9/17 : RUN apt-get -y --no-install-recommends install ca-certificates curl gnupg
 ---> Running in 9a8e0ee1c9da
Reading package lists...
Building dependency tree...
Reading state information...
Package ca-certificates is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

Package gnupg is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'ca-certificates' has no installation candidate
E: Unable to locate package curl
E: Package 'gnupg' has no installation candidate
E: Unable to locate package python3
The command '/bin/sh -c apt-get -y --no-install-recommends install ca-certificates curl gnupg python3' returned a non-zero code: 100
```

With `SHELL` directive enabled:

```
Step 13/18 : RUN   dpkg -i fail2ban.deb &>/dev/null
 ---> Running in eb24fdfc2111
The command '/bin/bash -o pipefail -c dpkg -i fail2ban.deb &>/dev/null' returned a non-zero code: 1
```

</details>

### Actual Behavior

With the actual Dockerfile, the error output is:

<details>
<summary>Output</summary>

```
Step 30/71 : SHELL ["/bin/bash", "-o", "pipefail", "-c"]
 ---> Using cache
 ---> 7e8f3851c231
Step 31/71 : RUN   apt-get -qq update &&   apt-get -y install apt-utils &>/dev/null &&   apt-get -y dist-upgrade >/dev/null &&   apt-get -y install postfix >/dev/null &&   apt-get -y --no-install-recommends install   altermime amavisd-new apt-transport-https arj binutils bzip2 bsd-mailx   ca-certificates cabextract clamav clamav-daemon cpio curl   dovecot-core dovecot-imapd dovecot-ldap dovecot-lmtpd   dovecot-managesieved dovecot-pop3d dovecot-sieve dovecot-solr   dumb-init   ed fetchmail file gamin gnupg gzip iproute2 iptables   locales logwatch lhasa libdate-manip-perl liblz4-tool   libmail-spf-perl libnet-dns-perl libsasl2-modules lrzip lzop   netcat-openbsd nomarch opendkim opendkim-tools opendmarc   pax pflogsumm postgrey p7zip-full postfix-ldap postfix-pcre   postfix-policyd-spf-python postsrsd pyzor   razor rpm2cpio rsyslog sasl2-bin spamassassin supervisor   unrar-free unzip whois xz-utils &&   gpg --keyserver ${FAIL2BAN_GPG_PUBLIC_KEY_SERVER}     --recv-keys ${FAIL2BAN_GPG_PUBLIC_KEY_ID} &>/dev/null &&   curl -Lkso fail2ban.deb ${FAIL2BAN_DEB_URL} &&   curl -Lkso fail2ban.deb.asc ${FAIL2BAN_DEB_ASC_URL} &&   FINGERPRINT="$(LANG=C gpg --verify   fail2ban.deb.asc fail2ban.deb 2>&1     | sed -n 's#Primary key fingerprint: \(.*\)#\1#p')" &&   if [[ -z ${FINGERPRINT} ]]; then     echo "ERROR: Invalid GPG signature!" 2>&1; exit 1; fi &&   if [[ ${FINGERPRINT} != "${FAIL2BAN_GPG_FINGERPRINT}" ]]; then     echo "ERROR: Wrong GPG fingerprint!" 2>&1; exit 1; fi &&   dpkg -i fail2ban.deb &>/dev/null &&   rm fail2ban.deb fail2ban.deb.asc &&   apt-get -qq autoremove &>/dev/null &&   apt-get -qq autoclean &&   apt-get -qq clean &&   rm -rf /var/lib/apt/lists/* &&   c_rehash &>/dev/null
 ---> Running in 5479bdfbd569

E: Sub-process /usr/bin/dpkg returned an error code (1)

The command '/bin/bash -o pipefail -c apt-get -qq update &&   apt-get -y install apt-utils &>/dev/null &&   apt-get -y dist-upgrade >/dev/null &&   apt-get -y install postfix >/dev/null &&   apt-get -y --no-install-recommends install   altermime amavisd-new apt-transport-https arj binutils bzip2 bsd-mailx   ca-certificates cabextract clamav clamav-daemon cpio curl   dovecot-core dovecot-imapd dovecot-ldap dovecot-lmtpd   dovecot-managesieved dovecot-pop3d dovecot-sieve dovecot-solr   dumb-init   ed fetchmail file gamin gnupg gzip iproute2 iptables   locales logwatch lhasa libdate-manip-perl liblz4-tool   libmail-spf-perl libnet-dns-perl libsasl2-modules lrzip lzop   netcat-openbsd nomarch opendkim opendkim-tools opendmarc   pax pflogsumm postgrey p7zip-full postfix-ldap postfix-pcre   postfix-policyd-spf-python postsrsd pyzor   razor rpm2cpio rsyslog sasl2-bin spamassassin supervisor   unrar-free unzip whois xz-utils &&   gpg --keyserver ${FAIL2BAN_GPG_PUBLIC_KEY_SERVER}     --recv-keys ${FAIL2BAN_GPG_PUBLIC_KEY_ID} &>/dev/null &&   curl -Lkso fail2ban.deb ${FAIL2BAN_DEB_URL} &&   curl -Lkso fail2ban.deb.asc ${FAIL2BAN_DEB_ASC_URL} &&   FINGERPRINT="$(LANG=C gpg --verify   fail2ban.deb.asc fail2ban.deb 2>&1     | sed -n 's#Primary key fingerprint: \(.*\)#\1#p')" &&   if [[ -z ${FINGERPRINT} ]]; then     echo "ERROR: Invalid GPG signature!" 2>&1; exit 1; fi &&   if [[ ${FINGERPRINT} != "${FAIL2BAN_GPG_FINGERPRINT}" ]]; then     echo "ERROR: Wrong GPG fingerprint!" 2>&1; exit 1; fi &&   dpkg -i fail2ban.deb &>/dev/null &&   rm fail2ban.deb fail2ban.deb.asc &&   apt-get -qq autoremove &>/dev/null &&   apt-get -qq autoclean &&   apt-get -qq clean &&   rm -rf /var/lib/apt/lists/* &&   c_rehash &>/dev/null' returned a non-zero code: 100
make: *** [Makefile:20: build] Error 100
```

When disabling the `SHELL` directive on the actual Dockerfile, running `make build` results in:

```
Step 30/70 : RUN   apt-get -qq update &&   apt-get -y install apt-utils &>/dev/null &&   apt-get -y dist-upgrade >/dev/null &&   apt-get -y install postfix >/dev/null &&   apt-get -y --no-install-recommends install   altermime amavisd-new apt-transport-https arj binutils bzip2 bsd-mailx   ca-certificates cabextract clamav clamav-daemon cpio curl   dovecot-core dovecot-imapd dovecot-ldap dovecot-lmtpd   dovecot-managesieved dovecot-pop3d dovecot-sieve dovecot-solr   dumb-init   ed fetchmail file gamin gnupg gzip iproute2 iptables   locales logwatch lhasa libdate-manip-perl liblz4-tool   libmail-spf-perl libnet-dns-perl libsasl2-modules lrzip lzop   netcat-openbsd nomarch opendkim opendkim-tools opendmarc   pax pflogsumm postgrey p7zip-full postfix-ldap postfix-pcre   postfix-policyd-spf-python postsrsd pyzor   razor rpm2cpio rsyslog sasl2-bin spamassassin supervisor   unrar-free unzip whois xz-utils &&   gpg --keyserver ${FAIL2BAN_GPG_PUBLIC_KEY_SERVER}     --recv-keys ${FAIL2BAN_GPG_PUBLIC_KEY_ID} &>/dev/null &&   curl -Lkso fail2ban.deb ${FAIL2BAN_DEB_URL} &&   curl -Lkso fail2ban.deb.asc ${FAIL2BAN_DEB_ASC_URL} &&   FINGERPRINT="$(LANG=C gpg --verify   fail2ban.deb.asc fail2ban.deb 2>&1     | sed -n 's#Primary key fingerprint: \(.*\)#\1#p')" &&   if [[ -z ${FINGERPRINT} ]]; then     echo "ERROR: Invalid GPG signature!" 2>&1; exit 1; fi &&   if [[ ${FINGERPRINT} != "${FAIL2BAN_GPG_FINGERPRINT}" ]]; then     echo "ERROR: Wrong GPG fingerprint!" 2>&1; exit 1; fi &&   dpkg -i fail2ban.deb &>/dev/null &&   rm fail2ban.deb fail2ban.deb.asc &&   apt-get -qq autoremove &>/dev/null &&   apt-get -qq autoclean &&   apt-get -qq clean &&   rm -rf /var/lib/apt/lists/* &&   c_rehash &>/dev/null
 ---> Running in 851a120aca3a
Removing intermediate container 851a120aca3a
 ---> 6d9f68aa9c37
Step 31/70 : RUN   echo '0 */6 * * * clamav /usr/bin/freshclam --quiet' >/etc/cron.d/clamav-freshclam &&   chmod 644 /etc/clamav/freshclam.conf &&   freshclam &&   sed -i 's/Foreground false/Foreground true/g' /etc/clamav/clamd.conf &&   mkdir /var/run/clamav &&   chown -R clamav:root /var/run/clamav &&   rm -rf /var/log/clamav/
 ---> Running in f16d39d5623b

/bin/sh: 1: cannot create /etc/cron.d/clamav-freshclam: Directory nonexistent

The command '/bin/sh -c echo '0 */6 * * * clamav /usr/bin/freshclam --quiet' >/etc/cron.d/clamav-freshclam &&   chmod 644 /etc/clamav/freshclam.conf &&   freshclam &&   sed -i 's/Foreground false/Foreground true/g' /etc/clamav/clamd.conf &&   mkdir /var/run/clamav &&   chown -R clamav:root /var/run/clamav &&   rm -rf /var/log/clamav/' returned a non-zero code: 2
make: *** [Makefile:20: build] Error 2
```

</details>

</details>

Which hints that something went wrong with prior package installation step..

### Expected Behavior

Successfully build the image locally.

## Your Environment

- OS: Vultr VPS, Ubuntu 21.04
- version: `master` branch
- available RAM: `2GB`
- Docker version: `20.10.7, build f0df350`

**Comment by polarathene** [comment]
**TL;DR:** Success of build will be dependent upon host `/etc/resolv.conf` contents.

That file is mounted by Docker and checked by the Debian `postfix` package post-install script (if `hostname` does not provide a satisfactory FQDN).

A workaround is a bit verbose. Best course of action may be to get upstream to fix their script..

---

The problem is caused by upstream debian package post install script. However this will only affect certain systems as it's dependent upon this logic (`/var/lib/dpkg/info/postfix.postinst` [extracted via instructions here](https://askubuntu.com/a/482936)):

```sh
myfqdn() {
    myhostname=$(hostname --fqdn 2>/dev/null || echo "")
    # If we don't have a name with a dot (which includes ""), then we have work.
    if [ $myhostname = ${myhostname%.*} ]; then
        # If it's empty, and we have /etc/hostname, try that.
        if [ -z $myhostname ] && [ -r /etc/hostname ]; then
            myhostname=$(cat /etc/hostname)
        fi
        # If we are still lacking a domain component, then try resolv.conf.
        if [ $myhostname = ${myhostname%.*} ]; then
            if [ -f /etc/resolv.conf ]; then
                # The resolver uses the last one found, and ignores the rest
                mydom=$(sed -n 's/^search[[:space:]]*\([^[:space:]]*\).*/\1/p;s/^domain[[:space:]]*\([^[:space:]]*\).*/\1/p' /etc/resolv.conf | tail -1)
                myhostname="$myhostname${mydom:+.$mydom}"
            else
                myhostname="$myhostname.UNKNOWN"
            fi
        fi
    fi
    echo $myhostname
}
```

In my case, the above logic sets the hostname with `myhostname="$myhostname${mydom:+.$mydom}"`. This adds one `.` between the `hostname` and the `domainname`, but it takes the `domainname` from `/etc/resolv.conf` from the first value for `search` property, this was `.`, resulting in a hostname/FQDN of `<container_id>..` which is invalid.

[Upstream Debian bug report](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=864609#60) classed as will not fix..

We could workaround this within the Dockerfile:

```bash
TMP_SEARCH="$(cat /etc/resolv.conf | grep ^search)" && \
cp /etc/resolv.conf /tmp/resolv.conf && \
sed -i 's|^search.*|search localhost|' /tmp/resolv.conf && \
cp /tmp/resolv.conf /etc/resolv.conf && \
apt-get -y install postfix >/dev/null && \
sed -i "s|^search.*|${TMP_SEARCH}|" /tmp/resolv.conf && \
cp /tmp/resolv.conf /etc/resolv.conf
```

Or patch the logic in the `postinst` script, which may not be any less verbose than above.. The above approach is due to [`sed` not being able to directly modify the `/etc/resolv.conf` file as Docker mounted it and it's inode cannot be changed](https://unix.stackexchange.com/questions/404189/find-and-sed-string-in-docker-got-error-device-or-resource-busy). Presumably, some lines can be dropped as I don't think it's necessary to restore the original state of that file.

Likewise, `hostnamectl` is not available and despite being root `hostname set-hostname whatever` also fails ([_confusingly stating you need to be root_](https://evolvingweb.ca/blog/changing-docker-hostnames-namespaces)). Docker does not let us control the hostname at build time apparently..

There's also [this alternative](https://stackoverflow.com/a/41810877), which temporarily replaces the `hostname` command:

```
RUN \
  mv /bin/hostname{,.bak} && \
  echo "echo myhost.local" > /bin/hostname && \
  chmod +x /bin/hostname && \
  apt-get -y install postfix >/dev/null && \
  mv /bin/hostname{.bak,}
```

That looks simpler? It worked well too. Alternatively, no fix and we document the bug will prevent building the image unless the host `/etc/resolv.conf` or `Dockerfile` is modified accordingly to workaround it.

I have reviewed my previous VPS instance from earlier this year (which I still have running), it is using Ubuntu 20.10 on Vultr, and has the `/etc/resolv.conf` of:

```
nameserver 127.0.0.53
options edns0 trust-ad
```

While the Ubuntu 21.04 has:

```
nameserver 127.0.0.53
options edns0 trust-ad
search .
```

I have not verified if this is a general change for newer Ubuntu releases, or if it's Vultr specific. If the former it will eventually affect more users, including our CI.

**Comment by polarathene** [comment]
**TL;DR:**

- Environment change that triggers the bug is introduced from a newer systemd version (v247) by a new default behaviour.
- I will not be reporting the bug upstream due to an antiquated bug tracking system. Someone from the Debian community has expressed some interest in handling that though :+1: 

---

I briefly looked into reporting the bug upstream, but [the process](https://www.debian.org/Bugs/Reporting) is email based:

- There's a `reportbug` package for debian users to use as a CLI tool but AFAIK requires local mail setup and in my case presumably a relay to my personal e-mail service as I don't think I can send via port 25.
- Alternatively there's a non-CLI version, but some "pseudo-headers" are required. That document is vague on if these are actual headers to add, or part of the message body? So I'm not sure if that can be done via gmail mobile or web app easily?

Neither are methods I want to bother with for reporting a bug, so I'll not be filing a report upstream.

---

### Additional information

Previous environment without the issue has systemd v246, new environment has systemd v247. The newer version introduces a change that will [insert a default `search .` line into `/etc/resolv.conf` if no domain is configured](https://github.com/systemd/systemd/blob/33f2de7b6487c7fe7f45fd9d8fc73cd98e0a53ac/NEWS#L1341).

Thus if `/etc/resolv.conf` is controlled by `systemd-resolved` (fairly common these days AFAIK?), a host workaround is to edit that service config, eg `/etc/systemd/resolved.conf`:

```shell
$ sed -i 's|^#Domains=.*|Domains=test|' /etc/systemd/resolved.conf
$ systemctl restart systemd-resolved

# Verify update
$ cat /etc/resolv.conf | grep ^search
search test
```

This resolves the issue without any modifications to the Dockerfile. It should mostly only be required for some contributors? Definitely the simplest workaround. After building locally, the user can revert the modification to `/etc/resolv.conf` or similar if necessary.

AFAIK, this config change will just prepend certain queries like the hostname against the search domain (eg `myvps.test`). [`.test` is an officially reserved TLD name](https://en.wikipedia.org/wiki/.test), so it shouldn't cause any surprises? 

Info on `search` functionality with `resolv.conf`:

- [resolved.conf `Domains`](https://www.freedesktop.org/software/systemd/man/systemd.network.html#Domains=):

    > "search domains" and are first used as search suffixes for extending single-label hostnames (hostnames containing no dots) to become fully qualified domain names (FQDNs). If a single-label hostname is resolved on this interface, each of the specified search domains are appended to it in turn, converting it into a fully qualified domain name, until one of them may be successfully resolved.

    > Both "search" and "routing-only" domains are used for routing of DNS queries: look-ups for hostnames ending in those domains (hence also single label names, if any "search domains" are listed), are routed to the DNS servers configured for this interface.

    > This setting is read by systemd-resolved.service(8). "Search domains" correspond to the domain and search entries in resolv.conf(5). Domain name routing has no equivalent in the traditional glibc API, which has no concept of domain name servers limited to a specific link.
- [resolv.conf `search`](https://man7.org/linux/man-pages/man5/resolv.conf.5.html).
- [Article describing the behaviour better](https://fedoramagazine.org/systemd-resolved-introduction-to-split-dns/).
- [More technical info on resolution behaviour](https://unix.stackexchange.com/a/482402).

---

I can add this as a note into the contributor docs for anyone who runs into build problems. If upstream fixes it, once the image has the updated package the docs can drop the note?

**Comment by aendeavor** [comment]
@polarathene Your proposal seems fine to me. I will go ahead and clean the Dockerfile from the `>/dev/null`‘s and close #2014 on the way too.

**Comment by polarathene** [comment]
Several docs PRs in the works atm, once that settles, this issue will finally get some docs and be resolved :sweat_smile:

**Comment by NorseGaud** [comment]
Hi, so wouldn't this impact anyone building this to use in their server (not using dockerhub)? Not just contributors? Shouldn't we be looking for a Dockerfile solution to this so that the often not looked at documentation wouldn't be a blocker for someone wanting to just jump right in?

**Comment by polarathene** [comment]
> Hi, so wouldn't this impact anyone building this to use in their server (not using dockerhub)? Not just contributors?

It affects you only if you use a distro with the mentioned version of systemd or newer where `systemd-resolved` is configuring local DNS resolution. I'd need to read over what I wrote above to detail it better, but if IIRC, I solved it by having a fallback response when the container tries to perform some query, `.test` is fine as no public DNS resolver is allowed to handle that.

The actual bug is upstream in Debian by the packager for Postfix. A correct fix would address that, or build Postfix without the Debian package (which runs some extra scripts at install time). Once Debian itself updates with the newer version of systemd, it should likewise run into the same problem and the package maintainer will probably address it then I guess..

**Comment by NorseGaud** [comment]
A great example of why having ourselves dependent on others for fixing bugs is not a good idea generally ;)

**Comment by polarathene** [comment]
> A great example of why having ourselves dependent on others for fixing bugs is not a good idea generally ;)

If that's a reference to my advocating for `watchexec` for `check-for-changes.sh`, it doesn't have merit when our own bash scripts ran into problems due to upstream changes, nor the maintenance woes the script itself had with previous versions and new maintainers trying to grok history of the logic and decisions / changes involved for debugging needs.

I'm reluctant towards your proposed workaround patch for this issue, but I don't expect much interaction with `hostname` beyond the Debian post-install script for the Postfix package within that scope for the image build.. and there isn't much better options to address it within the image build, beyond modifying the post-install script directly instead which is more effort.

**Comment by polarathene** [comment]
This was [resolved with the suggested Dockerfile workaround](https://github.com/docker-mailserver/docker-mailserver/pull/2468) instead of the documentation update I never got around to.
